pkg/analyzer/analyzer.go

```diff
@@ -41,46 +81,116 @@ func newVisitor(pass *analysis.Pass) func(node ast.Node) {
 			return
 		}
 
-		var outgoingFieldList *ast.FieldList
+		funcName := a.funcName(deferStmt.Call)
+		if funcName != "" && a.exclude.MatchesAny(funcName) {
+			return
+		}
 
+		var isFuncReturnType bool
 		switch v := deferStmt.Call.Fun.(type) {
-		case *ast.Ident: // function is named
-			outgoingFieldList = getFuncDeclResults(v)
 		case *ast.FuncLit: // function is anonymous
-			outgoingFieldList = getFuncLitResults(v)
+			isFuncReturnType = a.isFuncLitReturnType(v)
+		case *ast.Ident:
+			isFuncReturnType = a.isIdentReturnType(v)
+		case *ast.SelectorExpr:
+			isFuncReturnType = a.isSelExprReturnType(v)
 		default:
 			return
 		}
 
-		if outgoingFieldList == nil || outgoingFieldList.List == nil {
-			return
-		}
-
-		if len(outgoingFieldList.List) == 0 {
+		if !isFuncReturnType {
 			return
 		}
 
 		pass.Reportf(node.Pos(), "deferred call should not return anything")
 	}
 }
 
-func getFuncDeclResults(ident *ast.Ident) *ast.FieldList {
-	if ident.Obj == nil {
-		return nil
+func (a *analyzer) isIdentReturnType(ident *ast.Ident) bool {
```

Comment:
Same here in terms of naming, this function checks presence of returned values, not its type.